.github/workflows/test.yaml

```diff
@@ -0,0 +1,343 @@
+            ~/.rustup/settings.toml
+          key: ${{ runner.os }}-rustup-${{ hashFiles('rust-toolchain') }}
+      - name: Cache Cargo
+        uses: actions/cache@v2
+        with:
+          # Note that for reasons unknown, possibly related to symlinking, this fails if
+          # we try to cache ~/.cargo/ directly.
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+          key: |
+            ${{ runner.os }}-cargo-${{ hashFiles('rust-toolchain') }}-${{ hashFiles('src/rust/engine/Cargo.*') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-${{ hashFiles('rust-toolchain') }}-
+      # Note: Must run before get-engine-hash, which uses the virtualenv.
+      - name: Cache Pants Virtualenv
+        uses: actions/cache@v2
+        with:
+          path: |
+            build-support/virtualenvs
+          key: |
+            ${{ runner.os }}-pants-venv-${{ matrix.python-version }}-${{ hashFiles('pants/3rdparty/python/**', 'pants.toml') }}
+      - name: Get Engine Hash
+        id: get-engine-hash
+        run: |
+          echo "::set-output name=hash::$(./build-support/bin/rust/print_engine_hash.sh)"
+        shell: bash
+      - name: Cache Native Engine
+        uses: actions/cache@v2
+        with:
+          path: |
+            src/python/pants/engine/internals/native_engine.so
+            src/python/pants/engine/internals/native_engine.so.metadata
+          key: |
+            ${{ runner.os }}-engine-${{ steps.get-engine-hash.outputs.hash }}
+      - name: Set env vars
+        run: |
+          echo "PANTS_CONFIG_FILES=${{ github.workspace }}/pants.travis-ci.toml" >> ${GITHUB_ENV}
+      - name: Bootstrap Pants
+        run: |
+          ./pants --version
+      - name: Run smoke tests
+        run: |
+          ./pants help goals
+          ./pants list ::
+          ./pants roots
+          ./pants help targets
+      - name: Test and Lint Rust
+        # Note that running these in a separate job would require that job to rebuild
+        # the Rust code from scratch, since Cargo uses mtimes to invalidate, and those
+        # are not conserved across git clones. And since we have to build Rust code
+        # in this job anyway, and the tests only take a couple of minutes to run, we
+        # run them in the same job.
+        run: |
+          sudo apt-get install -y pkg-config fuse libfuse-dev
+          ./cargo clippy --all
+          # We pass --tests to skip doc tests because our generated protos contain invalid
+          # doc tests in their comments.
+          ./cargo test --all --tests -- --nocapture
+        if: "!contains(env.COMMIT_MESSAGE, '[ci skip-rust]')"
+      - name: Upload native_engine.so
+        uses: actions/upload-artifact@v2
+        with:
+          name: native_engine.so
+          path: |
+            src/python/pants/engine/internals/native_engine.so
+            src/python/pants/engine/internals/native_engine.so.metadata
```

Comment:
What's the reasoning behind both caching and manually uploading/downloading these?

Reply:
It may not be necessary. The thinking was that A) caches might be trimmed between jobs (GHA trims your caches in some unspecified way if the total for your org exceeds 5 GB) and B) resources are long-lived (90 days I believe), meaning given an engine fingerprint we can download a built engine for that fingerprint. That seems useful for publishing etc. Should allow us to get rid of our public S3 bucket hack?

Reply:
Actually it may not be that simple, we might need to know a workflow run id (or the SHA it ran at) in order to get its artifacts by name. So possibly we would need to put the engine fingerprint in the artifact name itself, so we can do this globally without knowing which workflow run created the artifact. Still, can punt on that for now.